zookeeper-api/src/main/java/org/apache/helix/zookeeper/util/HttpRoutingDataReader.java

```
@@ -0,0 +1,134 @@
+ * under the License.
+ */
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.jcabi.aspects.RetryOnFailure;
+import org.apache.helix.msdcommon.constant.MetadataStoreRoutingConstants;
+import org.apache.helix.msdcommon.datamodel.MetadataStoreRoutingData;
+import org.apache.helix.msdcommon.datamodel.TrieRoutingData;
+import org.apache.helix.msdcommon.exception.InvalidRoutingDataException;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.codehaus.jackson.map.ObjectMapper;
+
+
+public class HttpRoutingDataReader {
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private static volatile Map<String, List<String>> _rawRoutingData;
+  private static volatile MetadataStoreRoutingData _metadataStoreRoutingData;
+
+  /**
+   * This class is a Singleton.
+   */
+  private HttpRoutingDataReader() {
+  }
+
+  /**
+   * Fetches routing data from the data source via HTTP.
+   * @return a mapping from "metadata store realm addresses" to lists of
+   * "metadata store sharding keys", where the sharding keys in a value list all route to
+   * the realm address in the key disallows a meaningful mapping to be returned
+   */
+  @RetryOnFailure(attempts = 3, delay = 1, unit = TimeUnit.SECONDS)
+  public static Map<String, List<String>> getRawRoutingData()
+      throws IOException {
+    if (_rawRoutingData == null) {
+      synchronized (HttpRoutingDataReader.class) {
+        if (_rawRoutingData == null) {
+          String msdsEndpoint =
+              System.getProperty(MetadataStoreRoutingConstants.MSDS_SERVER_ENDPOINT_KEY);
```

I am considering setting this in constructor. At least we can move it out of synchronized block.

---

I agree - I want to pull this out of the block and fail early if it doesn't exist.